packages/dcos-launch/extra/launch/config.py

```diff
@@ -307,11 +311,14 @@ def calculate_instance_count(num_masters, num_private_agents, num_public_agents)
     return str(1 + int(num_masters) + int(num_private_agents) + int(num_public_agents))
 
 
-def validate_os_name(os_name, platform):
+def validate_os_name(os_name, platform, provider):
     # TODO DCOS-14033: [gen.internals] Source validate functions are global only
-    if platform == 'bare_cluster':
+    if provider != 'onprem':
+        # the os_name parameter only applies to homogenous clusters that need
+        # to be created by dcos-launch for an onprem install. In other cases,
+        # the deployment template is parameterized for OS selection
         return
-    elif platform == 'aws':
+    if platform == 'aws':
```

Comment:
That's a good question, but the issue comes down to reporting. When one of these validation checks fail, the error is reported and tied to each of the variables the validate function requires. So, if you were to implement that and provide and invalid configuration value for platform/provider and a valid value for os_name, os_name would show as incorrect with a message about onprem platform support. This would be rather confusing. The goal is to get the errors to be as specific and directed as possible.

The great thing about this implementation is that in ANY of these functions that have platform and provider as arguments will have `validate_onprem_platform_provider` run in so no other validate function needs to worry about if its inputs are valid or not

The way this is set up now is to simply pass on all other situations that are unknown to that validate function. This was necessary because all validation functions are global, which means we cannot conditionally add them.

TL;DR Usually you would be right, but because of how `gen` is implemented, it would make the reported error obfuscated. But also because how `gen` is implemented, your suggestion would not add much functionally